src/indexer.ts

```diff
@@ -175,57 +179,139 @@ export class Index<K extends IndexKeyType, T extends DocTypes, R extends DocFrag
     }
   }
 
-  async query(opts: QueryOpts<K> = {}): Promise<IndexRows<K, T, R>> {
-    this.logger.Debug().Msg("enter query");
-    await this.ready();
-    // this._resetIndex();
-    this.logger.Debug().Msg("post ready query");
-    await this._updateIndex();
-    this.logger.Debug().Msg("post _updateIndex query");
-    await this._hydrateIndex();
-    this.logger.Debug().Msg("post _hydrateIndex query");
-    if (!this.byKey.root) {
-      return await applyQuery<K, T, R>(this.crdt, { result: [] }, opts);
-    }
-    if (this.includeDocsDefault && opts.includeDocs === undefined) opts.includeDocs = true;
-    if (opts.range) {
-      const eRange = encodeRange(opts.range);
-      return await applyQuery<K, T, R>(this.crdt, await throwFalsy(this.byKey.root).range(eRange[0], eRange[1]), opts);
-    }
-    if (opts.key) {
-      const encodedKey = encodeKey(opts.key);
-      return await applyQuery<K, T, R>(this.crdt, await throwFalsy(this.byKey.root).get(encodedKey), opts);
-    }
-    if (Array.isArray(opts.keys)) {
-      const results = await Promise.all(
-        opts.keys.map(async (key: DocFragment) => {
-          const encodedKey = encodeKey(key);
-          return (await applyQuery<K, T, R>(this.crdt, await throwFalsy(this.byKey.root).get(encodedKey), opts)).rows;
-        }),
+  query(opts: QueryOpts<K> = {}, { waitFor }: { waitFor?: Promise<unknown> } = {}): QueryResponse<T> {
```

Comment:
this method should splitted in a shorter one.

Reply:
@mabels See 032906d. That better?

Reply:
Why better.

Reply:
Shorter functions, less nested scoping and all that.